docs/testing-guide/contract-testing.md

```diff
@@ -0,0 +1,190 @@
+```
+
+## `algopy.ARC4Contract`
+
+Classes prefixed with `algopy.ARC4Contract` are **required** to be instantiated withing test context. As part of instantiation, the test context will automatically create a matching `algopy.Application` object instance.
+
+Within the class implementation, methods decorated with `algopy.arc4.abimethod` and `algopy.arc4.baremethod` will automatically assemble an `algopy.gtxn.ApplicationCallTransaction` transaction to emulate the AVM application call. This behavior can be overriden by setting the transaction group manually as part of test setup, this is done via implicit invocation of `algopy_testing.context.any_application()` _value generator_ (refer to [APIs](../apis.md) for more details).
+
+```{testcode}
+from algopy import ARC4Contract, GlobalState, LocalState, UInt64, Txn, arc4
+from algopy_testing import algopy_testing_context
+import pytest
+from algopy import UInt64
+
+class SimpleVotingContract(ARC4Contract):
+    def __init__(self):
+        self.topic = GlobalState(UInt64(0), key="topic", description="Current voting topic ID")
+        self.votes = GlobalState(UInt64(0), key="votes", description="Total votes cast")
+        self.has_voted = LocalState(UInt64(0), key="voted", description="Whether an account has voted")
+
+    @arc4.abimethod(create="require")
+    def create(self, initial_topic: UInt64) -> None:
+        self.topic.value = initial_topic
+
+    @arc4.abimethod
+    def vote(self) -> UInt64:
+        assert self.has_voted[Txn.sender] == UInt64(0), "Account has already voted"
+        self.votes.value += UInt64(1)
+        self.has_voted[Txn.sender] = UInt64(1)
+        return self.votes.value
+
+    @arc4.abimethod(readonly=True)
+    def get_votes(self) -> UInt64:
+        return self.votes.value
+
+    @arc4.abimethod
+    def change_topic(self, new_topic: UInt64) -> None:
+        assert Txn.sender == Txn.application_id.creator_address(), "Only creator can change topic"
+        self.topic.value = new_topic
+        self.votes.value = UInt64(0)
+        # Reset all votes (this is simplified and not efficient for many users)
+        self.has_voted.delete()
+
+...
+
+@pytest.fixture
+def context():
+    with algopy_testing_context() as ctx:
+        yield ctx
+        ctx.reset()
+
+def test_simple_voting_contract(context):
+    # Arrange
+    initial_topic = UInt64(1)
+    contract = SimpleVotingContract()
+
+    # Act - Create the contract
+    contract.create(initial_topic)
+
+    # Assert - Check initial state
+    assert contract.topic.value == initial_topic
+    assert contract.votes.value == UInt64(0)
+
+    # Act - Vote
+    # note how `algopy.Txn` is temporarily overriden to emulate the transaction content
+    # the method `.vote()` is decorated with `algopy.arc4.abimethod`, which means it will assemble a transaction to emulate the AVM application call
+    with context.scoped_txn_fields(sender=context.default_creator):
```

Comment:
This snippet looks like it's using the old API, additionally no need for a scope if using the default sender